File: src/chain-observers/classifiers/witness-classifier.ts
import { CollectorClassifierBase } from "./collector-classifier-base";

export interface IWitness {
  owner: string;
  runningVersion: string;
  totalMissedBlocks: number;
  lastConfirmedBlockNum: number;
}

export interface IWitnessData {
  witnesses: Record<string, IWitness>;
}

export interface IWitnessCollectorOptions {
  witness: string;
}

export class WitnessClassifier extends CollectorClassifierBase<IWitnessData, void, void, IWitnessCollectorOptions> {
  public static forOptions(options: IWitnessCollectorOptions) {
    return {
      class: this, // Intentionally using `this` to refer to the class prototype itself later - even though it is not a class **instance**
      options
    };
  }
}

File: python/workerbee/chain_observers/classifiers/witness_classifier.py
from __future__ import annotations

from typing import Any

from ..classifier_results import WitnessClassifierData
from .collector_classifier_base import CollectorClassifierBase, TRegisterEvaluationContext


class WitnessClassifier(CollectorClassifierBase[WitnessClassifierData]):
    """IWitnessData:
      witnesses: dict[str, IWitness]

    IWitness:
      owner: str
      running_version: str
      total_missed_blocks: int
      last_confirmed_block_num: int
    """

    @classmethod
    def for_options(cls, options: dict[str, Any]) -> TRegisterEvaluationContext:
        return {"class": cls, "options": options}
